subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/configurer/EclipseModelAwareUniqueProjectNameProvider.java

```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.plugins.ide.internal.configurer;
+
+import org.gradle.api.Project;
+
+import java.util.List;
+
+public interface EclipseModelAwareUniqueProjectNameProvider {
```

Comment:
I did try that first, but then the service registry contained two implementations of UniqueProjectNameProvider which caused a conflict.

Reply:
Does it even need to be a service? Why don't we just create a new instance in the model builder's constructor?

Reply:
Why don't we just move instantiation of the concrete class to the constructor of the model provider? That's the only place when we use this class and we wouldn't need the separate interface.

Reply:
I'm currently stubbing the whole ReservedNameAwareUniqueNameProvider in EclipseModelBuilderTest. This stub is quite simple. If we construct the UniqueProjectNameResolver in the constructor of the ModelBuilder we'd have to stub the ProjectStateRegistry which might require quite a bit more stubbing logic(or at least knowledge of the internal workings of the DefaultEclipseModelAwareUniqueProjectNameProvider.

Reply:
It can be resolved with two constructors, one taking the name provider, and one that calls `this(     public EclipseModelBuilder(gradleProjectBuilder, services, new UniqueNameProvider())`. The first constructor should have `@VisibleForTesting` annotation. IIRC you could even make that private: https://touk.pl/blog/2012/07/12/private-fields-and-methods-are-not-private-in-groovy/